[Vending machine step1] 아이패드 앱

1Consumption opened the pull request:
* CustomStringConvertible 프로토콜을 채택하여 인스턴스를 문자열로 표현하였습니다.
* Date 인스턴스를 원하는 포맷으로 출력하기 위해 DateFormatter()를 사용하여 년월일을 String 반환하였습니다.

1Consumption committed:
프로젝트 생성

자판기 앱을 위한 프로젝트 생성

1Consumption committed:
VendingMachine 클래스 생성

1Consumption committed:
Beveage 클래스 생성

1Consumption committed:
Beveage 클래스에 속성 추가

1Consumption committed:
접근자 수정

1Consumption committed:
Beverage 클래스에 CustomStringConvertible 프로토콜 채용

인스턴스를 문자열로 표현하기 위함

1Consumption committed:
ViewController에 Beverage 인스턴스 출력

godrm commented on VendingMachineApp/VendingMachineApp/AppDelegate.swift:
```diff
@@ -0,0 +1,37 @@
+    func application(_ application: UIApplication, didDiscardSceneSessions sceneSessions: Set<UISceneSession>) {
+        // Called when the user discards a scene session.
+        // If any sessions were discarded while the application was not running, this will be called shortly after application:didFinishLaunchingWithOptions.
+        // Use this method to release any resources that were specific to the discarded scenes, as they will not return.
```
프로젝트 템플릿으로 생성되는 메소드나 주석도 사용하지 않는 부분은 과감하게 삭제하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Beverage.swift:
```diff
@@ -0,0 +1,33 @@
+extension Beverage: CustomStringConvertible{
+    var description: String {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyyMMdd"
```
날짜를 형식에 맞춰서 바꾸는 것은 음료수 객체가 할 일이 아닌 것 같습니다. 
어떤 객체에 포함하는 게 좋을 지 생각해보고 옮겨보세요.

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachine.swift:
```diff
@@ -0,0 +1,13 @@
+
+class VendinMachine{
+    
+}
```
아무런 기능이 없고 사용하지 않는 클래스나 파일은 커밋할 때 포함하지 않는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -0,0 +1,21 @@
+        print(Beverage(brand: "서울우유", capacity: 200, price: 1000, name: "날마다딸기우유", manufacturingDate: Date()))
+        print(Beverage(brand: "서울우유", capacity: 200, price: 1000, name: "날마다딸기우유", manufacturingDate: Date(timeIntervalSinceNow: -86400 * 5)))
+        print(Beverage(brand: "펩시", capacity: 350, price: 2000, name: "다이어트콜라", manufacturingDate: Date()))
+        print(Beverage(brand: "맥심", capacity: 400, price: 3000, name: "TOP아메리카노", manufacturingDate: Date(timeIntervalSinceNow: -86400 * 2)))
```
날짜도 계산하는 값을 하드코딩하지 말고 상수로 선언하고 코드로 표현해서 의미 있는 단어로 읽을 수 있도록 해보세요.

godrm replied:
이 부분이 개선이 되지 않아서 아쉽습니다. 매번 생성할 때마다 -86400 곱하고 더하는 것을 지양하세요.

godrm reviewed (commented):
간단한 내용입니다. 객체 역할과 책임에 대해 피드백 확인하고 수정해보세요.

1Consumption committed:
사용하지 않는 메소드 제거

1Consumption committed:
하드 코딩 지양

상수로 선언하여 가독성 높임

1Consumption committed:
Beverages 클래스 생성

Beverage 객체 배열을 프로퍼티로 가지고, 동작에 대한 메소드를 가지고 있음

1Consumption committed:
Beverage 클래스에 메소드 추가

addBeverage(beverage: Beverage) : Beverage 배열에 엘리먼트를 추가해 줌
forEachBeverages(transform: (Beverage) -> ()) : Beverage 배열이 동작을 수행할 수 있도록 동작을 매개변수로 넘겨줌

1Consumption committed:
VendingMachine 클래스 변경

Beverage 배열을 가지지 않고 Beverages 객체를 가지게 하여 추상화 함
convertDate(date: Date) -> String : Beverage 객체의 Date를 사용자가 원하는 포맷으로 문자열로 반환해 줌
printBeverages() : Beverage들을 출력하고, Date를 포맷에 맞게 출력해 줌

1Consumption committed:
VendingMachine 클래스 변경

insertBeverage(beverage: Beverage) : Beverages 객체에 Beverage를 추가할 수 있는 인터페이스를 만들어 줌

1Consumption committed:
ViewController에 VendingMachine 인스턴스 추가

VendingMachine에 Beverage 객체를 추가해주고 출력함.

1Consumption commented:
피드백 반영하여 수정하였습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Beverage.swift:
```diff
@@ -22,12 +22,14 @@ class Beverage {
+    
+    func date() -> Date {
+        return manufacturingDate
+    }
```
왜 이 속성만 갑자기 getter 메소드가 생긴걸까요? 
CustomStringConvertible()로 표현하기 어려웠나요?

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachine.swift:
```diff
@@ -8,6 +8,26 @@
+    func convertDate(date: Date) -> String{
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyyMMdd"
+        return dateFormatter.string(from: date)
```
날짜를 변화하는 것도 자판기가 해야 할 역할과 책임은 아닌 것 같습니다. 
다른 타입들을 생각해보세요. 날짜를 다루는 것은 날짜를 다루는 객체가 해야하지 않을까요?

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachine.swift:
```diff
@@ -8,6 +8,26 @@
+    
+    func printBeverages(){
+        beverages.forEachBeverages{
+            print("\($0), \(convertDate(date: $0.date()))")
```
앞으로 미션에서는 출력을 하기 위한 메소드를 만들지 않는 게 좋습니다. 대신 적절한 데이터 타입을 만들어서 전달할 수 있는지 고민해보세요.

godrm commented:
다음 단계를 진행하기 전에 개선해보세요.
